src/client/js/pie-img.js

```diff
@@ -16,27 +16,31 @@
  *  limitations under the License
  *
  */
- (function() {
-  class PieImg extends HTMLElement {
-    static get observedAttributes() {
-      return ['dough', 'filling', 'topping'];
-    }
-    constructor() {
-      super();
-      this.style.width = this.getAttribute('width');
-      this.style.height = this.getAttribute('height');
-    }
-    _getSvgUrl() {
-      const topping = this.getAttribute('topping');
-      const dough = this.getAttribute('dough');
-      const filling = this.getAttribute('filling');
-      return `/pieimg/pie-svg?topping=${topping}&filling=${filling}&dough=${dough}`;
-    }
-    attributeChangedCallback(name, oldVal, newVal) {
-      if (['topping', 'dough', 'filling'].includes(name)) {
-        this.style.background = `url(${this._getSvgUrl()}) no-repeat center/cover`;
-      }
+export default class PieImg extends HTMLElement {
+  static get observedAttributes() {
+    return ['dough', 'filling', 'topping'];
+  }
+
+  static define() {
+    window.customElements.define('pie-img', PieImg);
+  }
```

Comment:
Ah, sorry I wasn't very clear. It's not so much that we're rewriting it into a class method, it's that we're stopping it from happening as soon as the file gets imported. That is, if it lives outside of a function scope, the custom element definition will get executed as soon as `pie-img` gets imported by something (conceptually, of course, since we're bundling). By wrapping it in a class method, you give control to the importer, so that it can define the custom elements whenever it wants.

In other words, I was trying to avoid `window.customElements.define` being called as a side-effect, and trying to make it more explicit. Do you think it's better to make it simpler and leave it as a side-effect?

Reply:
I'm all in for removing side effect. But why not have the importer call window.customElements.define directly? Why hide it in a custom "define" method?

Reply:
Oh, I see! It was just so that the client code didn't have to know the name of the HTML element, but I guess that it's actually valuable for it to be able to choose that as well. I'll make that change.